### ISSUE TITLE
When using async uploads, S3Mock incorrectly stores unsigned, chunked uploads without removing meta (chunk, checksum) information.

### ISSUE THREAD

**Issue by vlbaluk** [adobe/S3Mock#1662]
S3Mock is not functioning as expected with S3TransferManager.

1.  File content on the S3 mock Docker container contains data and metadata, leading to incorrect downloads.
2. S3Mock always returns incorrect length for multipart downloads, too.

Experiment: the file content is "BBBBB"

- Async CRT client with multipart and S3TransferManager stores the following file in S3Mock docker container: 

![BBBBBAsyncAttachment](https://github.com/adobe/S3Mock/assets/3883627/9a3edb5f-a0b4-4ede-b281-249a11c55a21)

- Synchronous client stores correct content: 

![ffe7b8b720bd49a98902fdb950d5bc97](https://github.com/adobe/S3Mock/assets/3883627/9a8cc156-987a-4c13-a7fb-c7f7d7e0def4)

Code snippet: 
```
store.write("test data".getBytes(UTF_8));
 // under the hood, s3TransferManager.uploadFile(uploadFileRequest) with CRT async client


s3TransferManager.download(downloadFileRequest).completionFuture().join().result().response().contentLength() 
// download returns incorrect content, including additional lines at the beginning and at the end, as on the screen above.  
                
```

Checked with 3.4.0 and 2.11.0 docker versions.

_Originally posted by @vlbaluk in https://github.com/adobe/S3Mock/issues/1613#issuecomment-1953877622_

**Comment by afranken** [comment]
@vlbaluk I added an integration test using TransferManager that is working just fine, see linked PR.

**Comment by afranken** [comment]
@vlbaluk when I change the line to `.checksumAlgorithm(ChecksumAlgorithm.CRC32)` or remove it alltogether, the test still runs without problems.

**Comment by afranken** [comment]
I added checksum support in version 2.17.0:
https://github.com/adobe/S3Mock/releases/tag/2.17.0
maybe you tested with an older version?

Older versions would ignore the additional bytes the AWS SDK adds to the payload when asking for checksum validation of (mulitpart) uploads.
The old behaviour would match your screenshot.

**Comment by afranken** [comment]
see also #1123

**Comment by glennvdv** [comment]
Got the same problem using 3.5.1
Sample code
`    @Test
    void testPutAndGetObject() throws Exception {
        URL resource = Thread.currentThread().getContextClassLoader().getResource("jon.png");
        var uploadFile = new File(resource.toURI());
 s3AsyncClient.putObject(PutObjectRequest.builder().bucket("eojt").key(uploadFile.getName()).build(),AsyncRequestBody.fromFile(uploadFile)).get();

        var response =
                s3Client.getObject(
                        GetObjectRequest.builder().bucket("eojt").key(uploadFile.getName()).build());

        var uploadFileIs = Files.newInputStream(uploadFile.toPath());
        var uploadDigest = hexDigest(uploadFileIs);
        var downloadedDigest = hexDigest(response);
        uploadFileIs.close();
        response.close();

        Assertions.assertThat(uploadDigest).isEqualTo(downloadedDigest);
    }`
Test output:
`Expected :"dcd37a339ac2f037a7498b9fc63048bb"
Actual   :"930c76274807e15e0873cb30a9d0d012"`
In the file following content is added
`0
x-amz-checksum-crc32:ntdN8g==
`

**Comment by afranken** [comment]
@glennvdv how do you construct the async client?
I can't reproduce these errors locally.

**Comment by glennvdv** [comment]
Using spring boot and Spring Cloud for Amazon Web Services

**Comment by vlbaluk** [comment]
@afranken We used crtBuilder() for constructing client:

     final S3CrtAsyncClientBuilder s3AsyncClientBuilder = S3AsyncClient.crtBuilder()
                .maxConcurrency(100) 
                .minimumPartSizeInBytes(10 * MB)
                .thresholdInBytes(100 * MB)
                .region(Region.of(...))
                .credentialsProvider(...);

**Comment by afranken** [comment]
@glennvdv I meant the actual code you are using to construct the client. As I said, I can't reproduce the problem locally.
I'm using several different clients in the integration tests:
https://github.com/adobe/S3Mock/blob/main/integration-tests/src/test/kotlin/com/adobe/testing/s3mock/its/S3TestBase.kt#L115

@vlbaluk that looks almost exactly the same as the client I'm using in the integration tests:
https://github.com/adobe/S3Mock/blob/main/integration-tests/src/test/kotlin/com/adobe/testing/s3mock/its/S3TestBase.kt#L231

**Comment by glennvdv** [comment]
@afranken i let spring boot create the async client. From there source they do something like this

`	@Bean
	@ConditionalOnMissingBean
	S3AsyncClient s3AsyncClient(AwsCredentialsProvider credentialsProvider) {
		S3CrtAsyncClientBuilder builder = S3AsyncClient.crtBuilder().credentialsProvider(credentialsProvider)
				.region(this.awsClientBuilderConfigurer.resolveRegion(this.properties));
		Optional.ofNullable(this.awsProperties.getEndpoint()).ifPresent(builder::endpointOverride);
		Optional.ofNullable(this.properties.getEndpoint()).ifPresent(builder::endpointOverride);
		Optional.ofNullable(this.properties.getCrossRegionEnabled()).ifPresent(builder::crossRegionAccessEnabled);
		Optional.ofNullable(this.properties.getPathStyleAccessEnabled()).ifPresent(builder::forcePathStyle);

		if (this.properties.getCrt() != null) {
			S3CrtClientProperties crt = this.properties.getCrt();
			PropertyMapper propertyMapper = PropertyMapper.get();
			propertyMapper.from(crt::getMaxConcurrency).whenNonNull().to(builder::maxConcurrency);
			propertyMapper.from(crt::getTargetThroughputInGbps).whenNonNull().to(builder::targetThroughputInGbps);
			propertyMapper.from(crt::getMinimumPartSizeInBytes).whenNonNull().to(builder::minimumPartSizeInBytes);
			propertyMapper.from(crt::getInitialReadBufferSizeInBytes).whenNonNull()
					.to(builder::initialReadBufferSizeInBytes);
		}

		return builder.build();
	}`

**Comment by afranken** [comment]
after testing with different configurations of clients and upload files in different sizes, I may have found the problem:
Some clients decide dynamically whether to use chunked uploads or not unless explicitly configured.
Signing is also dynamically decided upon, unless explicitly configured.

We currently do not handle chunked, unsigned uploads correctly, either we cut off some of the chunks before persisting the bytes to disk, or we write the chunks together with their chunk boundaries to disk.
Either way, we persist the wrong data to disk and later return the wrong data.

**Comment by afranken** [comment]
uploading chunked, signed data works without issues, BTW.

**Comment by afranken** [comment]
@glennvdv  / @vlbaluk I just released 3.7.1 which now correctly handles unsigned, chunked uploads when using async http clients, as long as the uploaded files are below 16KB.

See #1818
